File: flask_application/issues/denial_of_service.py
import asyncio
from flask import Blueprint, request
import json

denial_of_service = Blueprint('denial_of_service', __name__)


class ValidationError(Exception):
    def __init__(self, message):
        super().__init__(message)


def validate_user(value):
    if (
        not isinstance(value, dict) or
        value is None or
        isinstance(value, list) or
        'name' not in value or
        not isinstance(value['name'], str) or
        len(value['name']) == 0
    ):
        raise ValidationError(f"Invalid User: {json.dumps(value)}")

    return {'name': value['name']}


def validate_user_array(value):
    if not isinstance(value, list):
        raise ValidationError(
            f"Expected user array received {json.dumps(value)}")

    return [validate_user(user) for user in value]


async def work(user):
    await asyncio.sleep(0.1)
    return user['name']


@denial_of_service.route('/denial-of-service', methods=['POST'])
async def list_users():
    try:
        users = validate_user_array(request.get_json())
        names = []

        for user in users:
            names.append(await work(user))

        return ', '.join(names)
    except ValidationError as err:
        return str(err), 400
    except Exception:
        return "An unexpected error occurred", 500

File: laravel-application/routes/issues/denial-of-service.php
<?php

use Illuminate\Support\Facades\Route;

class ValidationError extends Exception
{
  public function __construct($message)
  {
    parent::__construct($message);
  }
}

function validateUser($value)
{
  if (
    !is_array($value) ||
    $value === null ||
    !isset($value['name']) ||
    !is_string($value['name']) ||
    strlen($value['name']) === 0
  ) {
    throw new ValidationError('Invalid User: ' . json_encode($value));
  }

  return ['name' => $value['name']];
}

function validateUserArray($value)
{
  if (!is_array($value)) {
    throw new ValidationError('Expected user array received ' . json_encode($value));
  }

  return array_map('validateUser', $value);
}

function work($user)
{
  usleep(100000);
  return $user['name'];
}

Route::post('/', function () {
  try {
    $users = validateUserArray(request()->json()->all());
    $names = [];

    foreach ($users as $user) {
      $names[] = work($user);
    }

    return implode(', ', $names);
  } catch (Exception $e) {
    if ($e instanceof ValidationError) {
      return response($e->getMessage(), 400);
    } else {
      return response('An unexpected error occurred', 500);
    }
  }
});
